filewatcher/filewatcher.go

```diff
@@ -0,0 +1,88 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package filewatcher
+
+import (
+	"os"
+	"time"
+
+	"github.com/mitchellh/hashstructure"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+type FileWatcher struct {
+	files    []string
+	lastScan time.Time
+	lastHash uint64
+}
+
+func New(files ...string) *FileWatcher {
+	return &FileWatcher{
+		lastScan: time.Time{},
+		lastHash: 0,
+		files:    files,
+	}
+}
+
+// Scan scans all file paths and checks if the number of files or the modtime of the files changed
+// It returns the list of existing files, a boolean if anything in has changed and potential errors.
+// To detect changes not only mod time is compared but also the hash of the files list. This is required to
+// also detect files which were removed.
+// The modtime is compared based on second as normally mod-time is in seconds. If it is unclear if something changed
+// the method will return true for the changes. It is strongly recommend to call scan not more frequent then 1s.
+func (f *FileWatcher) Scan() ([]string, bool, error) {
+	updatedFiles := false
+	files := []string{}
+
+	lastScan := time.Now()
+	defer func() { f.lastScan = lastScan }()
+
+	for _, path := range f.files {
+		info, err := os.Stat(path)
+		if err != nil {
+			logp.Err("Error getting stats for file: %s", path)
+			continue
+		}
+
+		// Check if one of the files was changed recently
+		// File modification time can be in seconds. -1 + truncation is to cover for files which
+		// were created during this second.
+		// If the last scan was at 09:02:15.00001 it will pick up files which were modified also 09:02:14
+		// As this scan no necessarily picked up files form 09:02:14
```

Comment:
I cannot understand this line, do you know what this means?

Reply:
That (with the time math/truncate) this scan would also pick up the files created at 09:02:14, I'll try to re-word it.